crates/dbsp/src/trace/spine_async/merger.rs

```diff
@@ -0,0 +1,150 @@
+//! A compactor thread that merges the batches for the spine-fueled trace.
+
+use crate::storage::backend::metrics::{COMPACTION_DURATION, COMPACTION_SIZE, TOTAL_COMPACTIONS};
+use crate::storage::backend::StorageError as Error;
+use crate::trace::spine_async::BatchIdent;
+use crate::trace::{Batch, Filter, Merger};
+use crate::Runtime;
+use crossbeam::channel::{bounded, unbounded, Receiver, RecvError, Sender};
+use metrics::{counter, histogram};
+use std::marker::PhantomData;
+use std::sync::Arc;
+use std::time::Instant;
+
+pub(super) enum MergeOperation<B>
+where
+    B: Batch,
+{
+    Merge(
+        BatchIdent,
+        Vec<Arc<B>>,
+        Option<Filter<B::Key>>,
+        Option<Filter<B::Val>>,
+    ),
+}
+
+pub(super) enum MergeResult<B>
+where
+    B: Batch,
+{
+    MergeCompleted(Result<(BatchIdent, B), Error>),
+}
+
+pub(crate) struct BatchMerger<B>
+where
+    B: Batch,
+{
+    /// A handle to initiate merge operations on the merger thread.
+    sender: Sender<MergeOperation<B>>,
+    /// A handle to receive completed merge operations from the merger thread.
+    receiver: Receiver<MergeResult<B>>,
+    /// In progress merges.
+    _in_progress: Vec<(MergeOperation<B>, <B as Batch>::Merger)>,
+    /// To make sure we use the same batch type as the spine using this.
+    _phantom: PhantomData<B>,
+}
+
+impl<B> BatchMerger<B>
+where
+    B: Batch,
+{
+    /// We sample the queue length only once every `QUEUE_LENGTH_SAMPLE_RATE` merges.
+    const QUEUE_LENGTH_SAMPLE_RATE: usize = 1_000;
+
+    /// Size of the incoming merge queue.
+    const RX_QUEUE_SIZE: usize = 64;
+    /// How many concurrent merges we allow.
+    const CONCURRENT_MERGES: usize = 1;
+
+    pub(crate) fn new() -> Self {
+        let (new_merge_sender, new_merge_receiver) = bounded(Self::RX_QUEUE_SIZE);
+        let (completed_merge_sender, completed_merge_receiver) = unbounded();
+        let mut merge_cnt = 0;
+
+        Runtime::spawn_compactor(move || loop {
+            let op: Result<MergeOperation<B>, RecvError> = new_merge_receiver.recv();
+            merge_cnt += 1;
+            if merge_cnt % Self::QUEUE_LENGTH_SAMPLE_RATE == 0 {
```

Comment:
Some merges can take multiple seconds, so this won't sample very often if those are occurring (and if merges are taking a long time, that's when the queue is likely to grow long)